ccl/sqlccl: TestBackupRestoreResume failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/d9de03641bd289beb8945dbf828bb7a74efd5e5a

Parameters:
```
TAGS=
GOFLAGS=
```

Stress build found a failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=481427&tab=buildLog

```
I180117 07:27:57.839756 133350 ccl/storageccl/export.go:124  [n1,s1,r31/1:/{Table/51/1/9…-Max}] export [/Table/51/1/900,/Table/51/2)
I180117 07:27:57.840232 133071 ccl/storageccl/export.go:124  [n1,s1,r27/1:/Table/51/1/{500-600}] export [/Table/51/1/500,/Table/51/1/600)
I180117 07:27:57.840448 133284 ccl/storageccl/export.go:124  [n1,s1,r25/1:/Table/51/1/{300-400}] export [/Table/51/1/300,/Table/51/1/400)
I180117 07:27:57.842291 133201 ccl/storageccl/export.go:124  [n1,s1,r26/1:/Table/51/1/{400-500}] export [/Table/51/1/400,/Table/51/1/500)
I180117 07:27:57.843251 133269 ccl/storageccl/export.go:124  [n1,s1,r29/1:/Table/51/1/{700-800}] export [/Table/51/1/700,/Table/51/1/800)
I180117 07:27:57.846448 133362 ccl/storageccl/export.go:124  [n1,s1,r28/1:/Table/51/1/{600-700}] export [/Table/51/1/600,/Table/51/1/700)
I180117 07:27:57.846587 133364 ccl/storageccl/export.go:124  [n1,s1,r22/1:/Table/51{-/1/100}] export [/Table/51/1,/Table/51/1/100)
I180117 07:27:57.847311 133363 ccl/storageccl/export.go:124  [n1,s1,r30/1:/Table/51/1/{800-900}] export [/Table/51/1/800,/Table/51/1/900)
I180117 07:27:57.847536 133365 ccl/storageccl/export.go:124  [n1,s1,r23/1:/Table/51/1/{100-200}] export [/Table/51/1/100,/Table/51/1/200)
I180117 07:27:57.854942 133301 ccl/storageccl/export.go:124  [n1,s1,r24/1:/Table/51/1/{200-300}] export [/Table/51/1/200,/Table/51/1/300)
I180117 07:27:57.907684 132133 sql/event_log.go:115  [client=127.0.0.1:45344,user=root,n1] Event: "create_database", target: 52, info: {DatabaseName:restoredb Statement:CREATE DATABASE restoredb User:root}
I180117 07:27:57.910313 130401 storage/replica_command.go:831  [split,n1,s1,r31/1:/{Table/51/1/9…-Max}] initiating a split of this range at key /Table/52 [r32]
I180117 07:27:57.918951 130565 storage/replica_raftstorage.go:540  [raftsnapshot,n1,s1,r31/1:/Table/5{1/1/900-2}] generated Raft snapshot 655c25cd at index 15
E180117 07:27:57.922160 130565 storage/queue.go:668  [raftsnapshot,n1,s1,r31/1:/Table/5{1/1/900-2}] snapshot failed: (n3,s3):3: remote couldn't accept snapshot with error: [n3,s3],r31: cannot apply snapshot: snapshot intersects existing range [n3,s3,r30/3:/{Table/51/1/8…-Max}]
E180117 07:27:58.034736 130984 sql/jobs/registry.go:208  error while adopting jobs: current lease node_id:3 epoch:1  did not match expected lease node_id:1 
github.com/cockroachdb/cockroach/pkg/sql/jobs.(*Job).adopt.func1
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/jobs/jobs.go:472
github.com/cockroachdb/cockroach/pkg/sql/jobs.(*Job).update.func1
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/jobs/jobs.go:430
github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn.func1
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:508
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:753
github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:507
github.com/cockroachdb/cockroach/pkg/sql/jobs.(*Job).runInTxn
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/jobs/jobs.go:347
github.com/cockroachdb/cockroach/pkg/sql/jobs.(*Job).update
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/jobs/jobs.go:414
github.com/cockroachdb/cockroach/pkg/sql/jobs.(*Job).adopt
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/jobs/jobs.go:467
github.com/cockroachdb/cockroach/pkg/sql/jobs.(*Registry).maybeAdoptJob
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/jobs/registry.go:496
github.com/cockroachdb/cockroach/pkg/sql/jobs.(*Registry).Start.func2
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/jobs/registry.go:207
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:195
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:2337
unable to acquire lease
github.com/cockroachdb/cockroach/pkg/sql/jobs.(*Registry).maybeAdoptJob
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/jobs/registry.go:497
github.com/cockroachdb/cockroach/pkg/sql/jobs.(*Registry).Start.func2
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/jobs/registry.go:207
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:195
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:2337
I180117 07:27:58.251858 131071 storage/replica_proposal.go:196  [n2,s2,r22/2:/Table/51{-/1/100}] new range lease repl=(n2,s2):2 seq=2 start=1516174078.247625825,0 epo=1 pro=1516174078.247629103,0 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=1516174082.222899876,0 pro=1516174073.222922812,0
I180117 07:27:58.792781 130565 storage/replica_raftstorage.go:540  [raftsnapshot,n1,s1,r30/1:/Table/51/1/{800-900}] generated Raft snapshot 1c838646 at index 16
I180117 07:27:58.793615 130565 storage/store.go:3670  [raftsnapshot,n1,s1,r30/1:/Table/51/1/{800-900}] streamed snapshot to (n3,s3):3: kv pairs: 110, log entries: 6, rate-limit: 8.0 MiB/sec, 1ms
I180117 07:27:58.793938 134386 storage/replica_raftstorage.go:746  [n3,s3,r30/3:/{Table/51/1/8…-Max}] applying Raft snapshot at index 16 (id=1c838646, encoded size=15665, 1 rocksdb batches, 6 log entries)
I180117 07:27:58.794398 134386 storage/replica_raftstorage.go:752  [n3,s3,r30/3:/Table/51/1/{800-900}] applied Raft snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I180117 07:27:58.992752 130565 storage/replica_raftstorage.go:540  [raftsnapshot,n1,s1,r31/1:/Table/5{1/1/900-2}] generated Raft snapshot b2eb4236 at index 16
I180117 07:27:58.993497 130565 storage/store.go:3670  [raftsnapshot,n1,s1,r31/1:/Table/5{1/1/900-2}] streamed snapshot to (n3,s3):3: kv pairs: 110, log entries: 6, rate-limit: 8.0 MiB/sec, 1ms
I180117 07:27:58.993826 134423 storage/replica_raftstorage.go:746  [n3,s3,r31/3:{-}] applying Raft snapshot at index 16 (id=b2eb4236, encoded size=15597, 1 rocksdb batches, 6 log entries)
I180117 07:27:58.994486 134423 storage/replica_raftstorage.go:752  [n3,s3,r31/3:/Table/5{1/1/900-2}] applied Raft snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I180117 07:27:59.266740 131112 storage/replica_proposal.go:196  [n2,s2,r5/2:/System/ts{d-e}] new range lease repl=(n2,s2):2 seq=2 start=1516174079.264771232,0 epo=1 pro=1516174079.264774056,0 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=1516174082.222899876,0 pro=1516174073.222922812,0
I180117 07:28:00.268464 131393 storage/replica_proposal.go:196  [n3,s3,r9/2:/Table/1{2-3}] new range lease repl=(n3,s3):2 seq=2 start=1516174080.266372177,0 epo=1 pro=1516174080.266374920,0 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=1516174082.222899876,0 pro=1516174073.222922812,0
I180117 07:28:01.274174 131020 storage/replica_proposal.go:196  [n2,s2,r30/2:/Table/51/1/{800-900}] new range lease repl=(n2,s2):2 seq=2 start=1516174081.271565792,0 epo=1 pro=1516174081.271568861,0 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=1516174082.222899876,0 pro=1516174073.222922812,0
I180117 07:28:01.731582 130431 storage/replica_proposal.go:196  [n1,s1,r7/1:/Table/{SystemCon…-11}] new range lease repl=(n1,s1):1 seq=2 start=0.000000000,0 epo=1 pro=1516174081.730558729,0 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=1516174082.222899876,0 pro=1516174073.222922812,0
I180117 07:28:01.735717 130466 storage/replica_proposal.go:196  [n1,s1,r12/1:/Table/1{5-6}] new range lease repl=(n1,s1):1 seq=2 start=0.000000000,0 epo=1 pro=1516174081.735006968,0 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=1516174082.222899876,0 pro=1516174073.222922812,0
I180117 07:28:02.273422 130468 storage/replica_proposal.go:196  [n1,s1,r27/1:/Table/51/1/{500-600}] new range lease repl=(n1,s1):1 seq=2 start=0.000000000,0 epo=1 pro=1516174082.272494020,0 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=1516174082.222899876,0 pro=1516174073.222922812,0
I180117 07:28:02.569310 131321 storage/replica_proposal.go:196  [n3,s3,r28/3:/Table/51/1/{600-700}] new range lease repl=(n3,s3):3 seq=2 start=1516174082.222899876,1 epo=1 pro=1516174082.567269934,0 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=1516174082.222899876,0 pro=1516174073.222922812,0
I180117 07:28:03.224430 130080 server/status/runtime.go:219  [n1] runtime stats: 377 MiB RSS, 761 goroutines, 37 MiB/50 MiB/112 MiB GO alloc/idle/total, 165 MiB/219 MiB CGO alloc/total, 0.00cgo/sec, 0.00/0.00 %(u/s)time, 0.00 %gc (527x)
I180117 07:28:03.351403 131133 storage/replica_raftstorage.go:540  [raftsnapshot,n2,s2,r32/2:/{Table/52-Max}] generated Raft snapshot c0f08ce7 at index 11
I180117 07:28:03.351984 131133 storage/store.go:3670  [raftsnapshot,n2,s2,r32/2:/{Table/52-Max}] streamed snapshot to (n3,s3):3: kv pairs: 9, log entries: 1, rate-limit: 8.0 MiB/sec, 1ms
I180117 07:28:03.352426 135964 storage/replica_raftstorage.go:746  [n3,s3,r32/3:{-}] applying Raft snapshot at index 11 (id=c0f08ce7, encoded size=541, 1 rocksdb batches, 1 log entries)
I180117 07:28:03.352624 135964 storage/replica_raftstorage.go:752  [n3,s3,r32/3:/{Table/52-Max}] applied Raft snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I180117 07:28:03.374301 131351 storage/replica_proposal.go:196  [n3,s3,r32/3:/{Table/52-Max}] new range lease repl=(n3,s3):3 seq=2 start=1516174082.222899876,1 epo=1 pro=1516174083.373144211,0 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=1516174082.222899876,0 pro=1516174073.222922812,0
I180117 07:28:03.375810 135973 storage/replica_command.go:831  [n3,s3,r32/3:/{Table/52-Max}] initiating a split of this range at key /Table/53/1/500 [r41]
I180117 07:28:03.381044 131369 storage/replica_proposal.go:196  [n3,s3,r10/2:/Table/1{3-4}] new range lease repl=(n3,s3):2 seq=2 start=1516174082.222899876,1 epo=1 pro=1516174083.379740493,0 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=1516174082.222899876,0 pro=1516174073.222922812,0
I180117 07:28:03.385833 131384 storage/replica_proposal.go:196  [n3,s3,r32/3:/{Table/52-Max}] new range lease repl=(n3,s3):3 seq=2 start=1516174082.222899876,1 epo=1 pro=1516174083.373144211,0 following repl=(n3,s3):3 seq=2 start=1516174082.222899876,1 epo=1 pro=1516174083.373144211,0
I180117 07:28:03.386863 136067 storage/replica_command.go:831  [n3,s3,r41/3:/{Table/53/1/5…-Max}] initiating a split of this range at key /Table/53/1/700 [r42]
I180117 07:28:03.387217 136055 storage/replica_command.go:831  [n3,s3,r41/3:/{Table/53/1/5…-Max}] initiating a split of this range at key /Table/53/1/600 [r43]
W180117 07:28:03.435190 135600 storage/engine/mvcc.go:2222  [n3,s3,r41/3:/{Table/53/1/5…-Max}] unable to find value for /Local/Range/Table/53/1/500/RangeDescriptor @ 1516174083.386886530,0
I180117 07:28:03.441359 131292 storage/replica_proposal.go:196  [n3,s3,r41/3:/{Table/53/1/5…-Max}] new range lease repl=(n3,s3):3 seq=2 start=1516174082.222899876,1 epo=1 pro=1516174083.373144211,0 following repl=(n3,s3):3 seq=2 start=1516174082.222899876,1 epo=1 pro=1516174083.373144211,0
W180117 07:28:03.444127 135599 storage/replica.go:2175  [n3,s3,r41/3:/Table/53/1/{500-700}] context canceled before command queue: 1 QueryTxn
I180117 07:28:03.446966 135989 storage/replica_command.go:831  [n3,s3,r42/3:/{Table/53/1/7…-Max}] initiating a split of this range at key /Table/53/1/900 [r44]
I180117 07:28:03.447403 135993 storage/replica_command.go:831  [n3,s3,r42/3:/{Table/53/1/7…-Max}] initiating a split of this range at key /Table/53/1/800 [r45]
E180117 07:28:03.501941 136199 storage/replica.go:5709  [n3,s3,r41/3:/Table/53/1/{500-700}] stalling replica due to: range does not match splits: (/Table/53/1/500-/Table/53/1/600) + (/Table/53/1/600-/Max) != [n3,s3,r41/3:/Table/53/1/{500-700}]
W180117 07:28:03.502220 136055 internal/client/txn.go:540  [n3,s3,r41/3:/Table/53/1/{500-700}] failure aborting transaction: replica corruption (processed=true): range does not match splits: (/Table/53/1/500-/Table/53/1/600) + (/Table/53/1/600-/Max) != [n3,s3,r41/3:/Table/53/1/{500-700}]; abort caused by: replica corruption (processed=true): range does not match splits: (/Table/53/1/500-/Table/53/1/600) + (/Table/53/1/600-/Max) != [n3,s3,r41/3:/Table/53/1/{500-700}]
W180117 07:28:03.502417 135993 internal/client/txn.go:540  [n3,s3,r42/3:/{Table/53/1/7…-Max}] failure aborting transaction: kv/txn_coord_sender.go:433: cannot commit a read-only transaction; abort caused by: context canceled
W180117 07:28:03.503006 133861 internal/client/txn.go:540  [n3] failure aborting transaction: writing transaction timed out or ran on multiple coordinators; abort caused by: failed to send RPC: sending to all 1 replicas failed; last error: {<nil> rpc error: code = Canceled desc = context canceled}
I180117 07:28:03.605703 130895 server/status/runtime.go:219  [n2] runtime stats: 383 MiB RSS, 760 goroutines, 39 MiB/46 MiB/112 MiB GO alloc/idle/total, 173 MiB/229 MiB CGO alloc/total, 0.00cgo/sec, 0.00/0.00 %(u/s)time, 0.00 %gc (529x)
I180117 07:28:03.623932 131422 storage/replica_command.go:831  [split,n3,s3,r32/3:/Table/5{2-3/1/500}] initiating a split of this range at key /Table/53 [r46]
W180117 07:28:03.624900 136306 storage/engine/mvcc.go:2255  [n1,s1,r12/1:/Table/1{5-6}] unable to find value for /Table/15/2/"running"/2018-01-17T07:27:57.911622Z/314912531785809921/0 @ 1516174083.518944003,1 (<nil> vs {ID:4fb652c3-5412-4c9c-985a-6cdda33172af Isolation:SERIALIZABLE Key:[136] Epoch:1 Timestamp:1516174083.518944003,1 Priority:321337 Sequence:8 BatchIndex:0})
W180117 07:28:03.626849 136363 storage/engine/mvcc.go:2255  [n1,s1,r12/1:/Table/1{5-6}] unable to find value for /Table/15/2/"running"/2018-01-17T07:27:57.911622Z/314912531785809921/0 @ 1516174083.518944003,1 (<nil> vs {ID:4fb652c3-5412-4c9c-985a-6cdda33172af Isolation:SERIALIZABLE Key:[136] Epoch:1 Timestamp:1516174083.518944003,1 Priority:321337 Sequence:8 BatchIndex:0})
I180117 07:28:03.647865 131405 storage/replica_proposal.go:196  [n3,s3,r32/3:/Table/5{2-3/1/500}] new range lease repl=(n3,s3):3 seq=2 start=1516174082.222899876,1 epo=1 pro=1516174083.373144211,0 following repl=(n3,s3):3 seq=2 start=1516174082.222899876,1 epo=1 pro=1516174083.373144211,0
I180117 07:28:03.679817 130423 storage/replica_proposal.go:196  [n1,s1,r24/1:/Table/51/1/{200-300}] new range lease repl=(n1,s1):1 seq=2 start=0.000000000,0 epo=1 pro=1516174083.678077312,0 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=1516174082.222899876,0 pro=1516174073.222922812,0
I180117 07:28:03.688872 130944 server/status/runtime.go:219  [n3] runtime stats: 383 MiB RSS, 751 goroutines, 34 MiB/48 MiB/112 MiB GO alloc/idle/total, 174 MiB/230 MiB CGO alloc/total, 0.00cgo/sec, 0.00/0.00 %(u/s)time, 0.00 %gc (529x)
I180117 07:28:03.879641 130467 storage/replica_proposal.go:196  [n1,s1,r26/1:/Table/51/1/{400-500}] new range lease repl=(n1,s1):1 seq=2 start=0.000000000,0 epo=1 pro=1516174083.878769676,0 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=1516174082.222899876,0 pro=1516174073.222922812,0
W180117 07:28:04.178657 131196 storage/raft_transport.go:465  [n1] raft transport stream to node 3 failed: rpc error: code = Unavailable desc = transport is closing
W180117 07:28:04.178689 131524 storage/raft_transport.go:465  [n1] raft transport stream to node 2 failed: rpc error: code = Unavailable desc = transport is closing
W180117 07:28:04.179165 131506 storage/raft_transport.go:465  [n3] raft transport stream to node 1 failed: rpc error: code = Unavailable desc = transport is closing
W180117 07:28:04.179212 131874 storage/raft_transport.go:465  [n3] raft transport stream to node 2 failed: rpc error: code = FailedPrecondition desc = grpc: the client connection is closing
W180117 07:28:04.179551 131529 storage/raft_transport.go:465  [n2] raft transport stream to node 1 failed: rpc error: code = FailedPrecondition desc = grpc: the client connection is closing
W180117 07:28:04.179684 131829 storage/raft_transport.go:465  [n2] raft transport stream to node 3 failed: rpc error: code = FailedPrecondition desc = grpc: the client connection is closing
E180117 07:28:04.180149 131273 sql/jobs/registry.go:208  error while adopting jobs: node unavailable; try another peer
```

--- a/c-deps/libroach/db.cc
+++ b/c-deps/libroach/db.cc
@@ -2207,10 +2207,8 @@ DBIterator* DBNewTimeBoundIter(DBEngine* db, DBTimestamp min_ts, DBTimestamp max
       return true;
     }
     // If the timestamp range of the table overlaps with the timestamp range we
-    // want to iterate, the table might contain timestamps we care about. For
-    // consistency with engineccl.MVCCIncrementalIterator, the min_ts bound is
-    // exclusive, but the max_ts bound is inclusive.
-    return max.compare(tbl_min->second) >= 0 && min.compare(tbl_max->second) < 0;
+    // want to iterate, the table might contain timestamps we care about.
+    return max.compare(tbl_min->second) >= 0 && min.compare(tbl_max->second) <= 0;
   };
   return db->NewIter(&opts);
 }

--- a/pkg/ccl/storageccl/engineccl/mvcc.go
+++ b/pkg/ccl/storageccl/engineccl/mvcc.go
@@ -62,7 +62,11 @@ func NewMVCCIncrementalIterator(
 	e engine.Reader, startTime, endTime hlc.Timestamp,
 ) *MVCCIncrementalIterator {
 	return &MVCCIncrementalIterator{
-		iter:      e.NewTimeBoundIterator(startTime, endTime),
+		// The call to startTime.Next() converts our half-open (start, end] time
+		// interval into the fully-inclusive [start, end] interval that
+		// NewTimeBoundIterator expects. This is strictly a performance
+		// optimization; omitting the call would still return correct results.
+		iter:      e.NewTimeBoundIterator(startTime.Next(), endTime),
 		startTime: startTime,
 		endTime:   endTime,
 	}

--- a/pkg/roachpb/data.go
+++ b/pkg/roachpb/data.go
@@ -869,9 +869,9 @@ func (t *Transaction) BumpEpoch() {
 	t.Epoch++
 }
 
-// TimeBounds returns the start and end timestamps which inclusively
-// cover all intents which were written as part of this transaction.
-func (t *Transaction) TimeBounds() (hlc.Timestamp, hlc.Timestamp) {
+// InclusiveTimeBounds returns start and end timestamps such that all intents written as
+// part of this transaction have a timestamp in the interval [start, end].
+func (t *Transaction) InclusiveTimeBounds() (hlc.Timestamp, hlc.Timestamp) {
 	min := t.OrigTimestamp
 	max := t.Timestamp
 	if t.Epoch != 0 {
@@ -880,7 +880,7 @@ func (t *Transaction) TimeBounds() (hlc.Timestamp, hlc.Timestamp) {
 		}
 		min = t.EpochZeroTimestamp
 	}
-	return min, max.Next() // Next() makes the end of the interval closed
+	return min, max
 }
 
 // Update ratchets priority, timestamp and original timestamp values (among

--- a/pkg/roachpb/data_test.go
+++ b/pkg/roachpb/data_test.go
@@ -335,21 +335,21 @@ func TestTransactionBumpEpoch(t *testing.T) {
 	}
 }
 
-func TestTransactionTimeBounds(t *testing.T) {
+func TestTransactionInclusiveTimeBounds(t *testing.T) {
 	verify := func(txn Transaction, expMin, expMax hlc.Timestamp) {
-		if min, max := txn.TimeBounds(); min != expMin || max != expMax {
+		if min, max := txn.InclusiveTimeBounds(); min != expMin || max != expMax {
 			t.Errorf("expected (%s-%s); got (%s-%s)", expMin, expMax, min, max)
 		}
 	}
 	origNow := makeTS(1, 1)
 	txn := MakeTransaction("test", Key("a"), 1, enginepb.SERIALIZABLE, origNow, 0)
-	verify(txn, origNow, origNow.Next())
+	verify(txn, origNow, origNow)
 	txn.Timestamp.Forward(makeTS(1, 2))
-	verify(txn, origNow, makeTS(1, 3))
+	verify(txn, origNow, makeTS(1, 2))
 	txn.Restart(1, 1, makeTS(2, 1))
-	verify(txn, origNow, makeTS(2, 2))
+	verify(txn, origNow, makeTS(2, 1))
 	txn.Timestamp.Forward(makeTS(3, 1))
-	verify(txn, origNow, makeTS(3, 2))
+	verify(txn, origNow, makeTS(3, 1))
 }
 
 // TestTransactionObservedTimestamp verifies that txn.{Get,Update}ObservedTimestamp work as

--- a/pkg/storage/batcheval/cmd_resolve_intent_range.go
+++ b/pkg/storage/batcheval/cmd_resolve_intent_range.go
@@ -56,7 +56,7 @@ func ResolveIntentRange(
 	// Use a time-bounded iterator as an optimization if indicated.
 	var iterAndBuf engine.IterAndBuf
 	if args.MinTimestamp != (hlc.Timestamp{}) {
-		iter := batch.NewTimeBoundIterator(args.MinTimestamp, args.IntentTxn.Timestamp.Next())
+		iter := batch.NewTimeBoundIterator(args.MinTimestamp, args.IntentTxn.Timestamp)
 		iterAndBuf = engine.GetBufUsingIter(iter)
 	} else {
 		iterAndBuf = engine.GetIterAndBuf(batch)

--- a/pkg/storage/engine/engine.go
+++ b/pkg/storage/engine/engine.go
@@ -151,7 +151,12 @@ type Reader interface {
 	NewIterator(prefix bool) Iterator
 	// NewTimeBoundIterator is like NewIterator, but the underlying iterator will
 	// efficiently skip over SSTs that contain no MVCC keys in the time range
-	// (start, end].
+	// [start, end].
+	//
+	// Note that time-bound iterators are strictly a performance optimization, and
+	// will frequently return keys outside of the [start, end] time range. If you
+	// must guarantee that you never see a key outside of the time bounds, perform
+	// your own filtering.
 	NewTimeBoundIterator(start, end hlc.Timestamp) Iterator
 }
 

--- a/pkg/storage/gc_queue_test.go
+++ b/pkg/storage/gc_queue_test.go
@@ -739,12 +739,7 @@ func TestGCQueueTransactionTable(t *testing.T) {
 		}
 		// Set a high Timestamp to make sure it does not matter. Only
 		// OrigTimestamp (and heartbeat) are used for GC decisions.
-		// Note that we can't simply use hlc.MaxTimestamp because when
-		// we attempt to resolve a range of intents, the method
-		// Transaction.TimeBound() makes a call to hlc.Timestamp.Next()
-		// in order to create a closed interval on the range for a time
-		// bounded iterator to discover the new intents.
-		txn.Timestamp.Forward(hlc.MaxTimestamp.Prev())
+		txn.Timestamp.Forward(hlc.MaxTimestamp)
 		txns[strKey] = *txn
 		for _, addrKey := range []roachpb.Key{baseKey, outsideKey} {
 			key := keys.TransactionKey(addrKey, txn.ID)

--- a/pkg/storage/intent_resolver.go
+++ b/pkg/storage/intent_resolver.go
@@ -491,7 +491,7 @@ func (ir *intentResolver) cleanupFinishedTxnIntents(
 	ctx context.Context, txn *roachpb.Transaction, intents []roachpb.Intent, now hlc.Timestamp,
 ) error {
 	// Resolve intents.
-	min, _ := txn.TimeBounds()
+	min, _ := txn.InclusiveTimeBounds()
 	opts := ResolveOptions{Wait: true, Poison: false, MinTimestamp: min}
 	if err := ir.resolveIntents(ctx, intents, opts); err != nil {
 		return errors.Wrapf(err, "failed to resolve intents")

--- a/pkg/storage/replica_command.go
+++ b/pkg/storage/replica_command.go
@@ -487,7 +487,7 @@ func resolveLocalIntents(
 		desc = &mergeTrigger.LeftDesc
 	}
 
-	min, max := txn.TimeBounds()
+	min, max := txn.InclusiveTimeBounds()
 	iter := batch.NewTimeBoundIterator(min, max)
 	iterAndBuf := engine.GetBufUsingIter(iter)
 	defer iterAndBuf.Cleanup()

